src/Confluent.Kafka/Producer.cs

```diff
@@ -634,16 +636,46 @@ internal Producer(ProducerBuilder<TKey, TValue> builder)
             }
 
             var configHandle = SafeConfigHandle.Create();
+            IntPtr configPtr = configHandle.DangerousGetHandle();
+
+            Func<PartitionerDelegate, SafeTopicConfigHandle> createTopicConfigWithPartitioner = (partitioner) => {
+                var topicConfigHandle = SafeTopicConfigHandle.Create();
+                IntPtr topicConfigPtr = topicConfigHandle.DangerousGetHandle();
+                Librdkafka.topic_conf_set_partitioner_cb(topicConfigPtr,
+                    (IntPtr rkt, IntPtr keydata, UIntPtr keylen, int partition_cnt, IntPtr rkt_opaque, IntPtr msg_opaque) =>
+                    {
+                        unsafe
+                        {
+                            var topicNamePtr = Librdkafka.topic_name(rkt);
+                            var topic = Util.Marshal.PtrToStringUTF8(topicNamePtr);
+                            var keyIsNull = keydata == IntPtr.Zero;
+                            var keyBytes = keyIsNull
+                                ? ReadOnlySpan<byte>.Empty
+                                : new ReadOnlySpan<byte>(keydata.ToPointer(), (int)keylen);
+                            return partitioner(topic, partition_cnt, keyBytes, keyIsNull);
+                        }
+                    });
+                return topicConfigHandle;
+            };
+
+            // Configure the default custom partitioner, if specified, by creating a new
+            // default topic config object. It's important that this occurs before the
+            // other properties are set, since if any of those were fallthru properties,
+            // their values would get clobbered by the default ones in the topic config
+            // object we create here.
+            if (defaultPartitioner != null)
+            {
+                var defaultTopicConfigHandle = createTopicConfigWithPartitioner(defaultPartitioner);
```

Comment:
Any time you specify a topic_conf object to topic_new() you will loose whatever default topic config is accumulated on the global config, which is not what we want, so I think what we want to do is get the default_topic_config of the rk and then apply the partitioner to that and pass that object to topic_new(), both in the topic-specific and default case.
See `rd_kafka_default_topic_conf_dup()`

Reply:
thanks for picking up on this ...